[css-anchor-position-1] Evaluate position-try-fallbacks with try-tactic

anttijk opened the pull request:
#### d86d00141e75e84612b84cd452cc15f54a7cecb2
<pre>
[css-anchor-position-1] Evaluate position-try-fallbacks with try-tactic
<a href="https://bugs.webkit.org/show_bug.cgi?id=288287">https://bugs.webkit.org/show_bug.cgi?id=288287</a>
<a href="https://rdar.apple.com/145359529">rdar://145359529</a>

Reviewed by Alan Baradlay.

Implement basic avaluation support for position-try-fallbacks property.
Only &lt;try-tactic&gt; values flip-inline and flip-block are supported by this patch.

<a href="https://drafts.csswg.org/css-anchor-position-1/#position-try-fallbacks">https://drafts.csswg.org/css-anchor-position-1/#position-try-fallbacks</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic-anchor-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic-anchor.html: Added.

Add a basic test. Almost all existing WPTs for &lt;try-tactic&gt; also use @position-try rules.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::interleavedLayout):

Add a simpified layout entry point for interleaving.

* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::willBeDestroyed):
(WebCore::RenderBox::styleWillChange):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::registerPositionTryBox):
(WebCore::RenderView::unregisterPositionTryBox):

Track boxes with position-try for invalidation.

* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::computeInsetValue):
(WebCore::Style::AnchorPositionEvaluator::evaluate):

Flip the sides used in anchor();

(WebCore::Style::flipHorizontal):
(WebCore::Style::flipVertical):
(WebCore::Style::AnchorPositionEvaluator::resolvePositionTryFallbackProperty):

Property flip helper.

(WebCore::Style::AnchorPositionEvaluator::overflowsContainingBlock):

Helper to test for overlow.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::Builder):
(WebCore::Style::Builder::applyProperty):

Flip the property as needed when building position option styles.

* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::BuilderState):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::positionTryFallback const):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForLayoutDependencies):
(WebCore::Style::Scope::invalidateForPositionTryFallbacks):

Invalidation support.

* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):

If we have position options return the next option instead of computing the style.

(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolveStartingStyle const):
(WebCore::Style::TreeResolver::resolveAfterChangeStyleForNonAnimated const):
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext const):

Add argument and rename to use in fallback resolution.

(WebCore::Style::TreeResolver::resolveComposedTree):
(WebCore::Style::TreeResolver::resolve):

Invalidate for interleaving if we have positions options we haven&apos;t chosen yet.

(WebCore::Style::TreeResolver::resolveElement):

Generate position options from position-try-fallbacks after resolving the style.

(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):

Go through position-try-fallbacks and generate a position option for each as needed.

(WebCore::Style::TreeResolver::generatePositionOption):

Spin the style builder to generate a position option.

(WebCore::Style::TreeResolver::tryChoosePositionOption):

See if the current style overflows the containing block and try the next option if it does.

* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: <a href="https://commits.webkit.org/290891@main">https://commits.webkit.org/290891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/793c9a5f8df595c24a4917d368221a4ef69d9f2e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91405 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96373 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42095 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93455 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11314 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19272 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70179 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27701 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94406 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8619 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82788 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50505 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/373 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41263 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78705 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/380 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98376 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18567 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18822 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78626 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78406 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19390 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22928 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18565 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23843 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18277 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21736 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20043 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -6112,4 +6119,10 @@ bool RenderBox::overflowChangesMayAffectLayout() const
     return !canUseOverlayScrollbars() && (style().overflowY() == Overflow::Auto || style().overflowX() == Overflow::Auto);
 }
 
+bool RenderBox::overflowsContainingBlock() const
```
This name reads like a generic function but it applies specific box edges and it's hard to tell at what point it is okay to call (maybe turn it into a static helper where these constraints are implicitly met?). don't feel too strongly but it seems like a function that is easy to misuse.

anttijk replied:
Made it a standalone function in AnchorPositiionEvaluator

anttijk committed:
[css-anchor-position-1] Evaluate position-try-fallbacks with try-tactic
https://bugs.webkit.org/show_bug.cgi?id=288287
rdar://145359529

Reviewed by Alan Baradlay.

Implement basic avaluation support for position-try-fallbacks property.
Only <try-tactic> values flip-inline and flip-block are supported by this patch.

https://drafts.csswg.org/css-anchor-position-1/#position-try-fallbacks

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic-anchor-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-basic-anchor.html: Added.

Add a basic test. Almost all existing WPTs for <try-tactic> also use @position-try rules.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolveStyle):
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::interleavedLayout):

Add a simpified layout entry point for interleaving.

* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::willBeDestroyed):
(WebCore::RenderBox::styleWillChange):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::registerPositionTryBox):
(WebCore::RenderView::unregisterPositionTryBox):

Track boxes with position-try for invalidation.

* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::computeInsetValue):
(WebCore::Style::AnchorPositionEvaluator::evaluate):

Flip the sides used in anchor();

(WebCore::Style::flipHorizontal):
(WebCore::Style::flipVertical):
(WebCore::Style::AnchorPositionEvaluator::resolvePositionTryFallbackProperty):

Property flip helper.

(WebCore::Style::AnchorPositionEvaluator::overflowsContainingBlock):

Helper to test for overlow.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::Builder):
(WebCore::Style::Builder::applyProperty):

Flip the property as needed when building position option styles.

* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::BuilderState):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::positionTryFallback const):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::invalidateForLayoutDependencies):
(WebCore::Style::Scope::invalidateForPositionTryFallbacks):

Invalidation support.

* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::styleForStyleable):

If we have position options return the next option instead of computing the style.

(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolveStartingStyle const):
(WebCore::Style::TreeResolver::resolveAfterChangeStyleForNonAnimated const):
(WebCore::Style::TreeResolver::resolveAgainInDifferentContext const):

Add argument and rename to use in fallback resolution.

(WebCore::Style::TreeResolver::resolveComposedTree):
(WebCore::Style::TreeResolver::resolve):

Invalidate for interleaving if we have positions options we haven't chosen yet.

(WebCore::Style::TreeResolver::resolveElement):

Generate position options from position-try-fallbacks after resolving the style.

(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):

Go through position-try-fallbacks and generate a position option for each as needed.

(WebCore::Style::TreeResolver::generatePositionOption):

Spin the style builder to generate a position option.

(WebCore::Style::TreeResolver::tryChoosePositionOption):

See if the current style overflows the containing block and try the next option if it does.

* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: https://commits.webkit.org/290891@main
